Add config check automation tool for wolfSSL configure

gojimmypi opened the pull request:
# Description

The wolfSSL `./configure` command has an extraordinary number of [options and parameters](https://www.wolfssl.com/documentation/manuals/wolfssl/chapter02.html) that select target platforms, fine tune installation,  optimize performance, select various features, and more. Managing and experimenting with the options can be a daunting experience, particularly for the new user.

This tool allows options to be specified in a more readable [command file](https://github.com/gojimmypi/wolfssl/blob/ConfigCheck/scripts/config_check/cmd.txt) with comments. The output of the [refresh script](https://github.com/gojimmypi/wolfssl/blob/ConfigCheck/scripts/config_check/refresh.sh) breaks out the resultant enabled and disabled features into more easily viewed summary files.

See the [README.md](https://github.com/gojimmypi/wolfssl/blob/ConfigCheck/scripts/config_check/README.md) for full details.

Fixes zd# n/a

# Testing

How did you test?

There's no change to the underlying wolfSSL library. This is only a configuration assistance tool.

Run the from the `config_check` directory:

```bash
cd ./scripts/config_check
./refresh.sh
```

Observe files created in the  `config_check` directory.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

dgarske commented:
@jpbland1 is going to review and give this a try.

kaleb-himes commented:
retest this please.

jpbland1 commented on config_check/README.md:
```diff
@@ -0,0 +1,89 @@
+The multi-line `./configure` command with all parameters on subsequent lines is stripped of comments and everything placed on a single line statement when executed.
+
+Upon execution, the entire output is set to a file called [output.txt](./output.txt). Additionally, the enabled/disabled features (those items with an asterisk in the output and the word "yes" or "no") 
+are separated and stored in the respective [Enabled-Features.txt](Enabled-Features.txt) and [Enabled-Features.txt](Enabled-Features.txt) files.
```
should be `Disabled-Features.txt`

gojimmypi replied:
good catch. updated.

jpbland1 commented on config_check/refresh.sh:
```diff
@@ -0,0 +1,95 @@
+echo ""
+
+# get the contents of the command file, trimming all text after the # character
+WOLFSSL_CMD="$(cat $WOLFSSL_CMD_FILE | cut -d'#' -f1)"
```
Is error checking important here? when I comment out `./configure` in cmd.txt it doesn't work but it can't really do any damage and just leads to my terminal telling me command not found followed by all the text below

gojimmypi replied:
I added some error checking and comments in the `cmd.txt` file. There's not much difference other than not doing the file operations after the `./configure` command.

jpbland1 reviewed (changes requested):
made some comments, tested it and it works on my setup

gojimmypi commented on config_check/refresh.sh:
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
```
erg, that's a recurring problem with my commits as WSL doesn't seem to care about that. I believe the `+x` flag should be applied now.

jpbland1 replied:
can you check this on a VM? I'm on arch using bash and it's still not executable. everything else seems to be working

gojimmypi replied:
That's pretty weird, as I'm certain that I set the flag in WSL. According to WSL, these are the permissions:

![image](https://user-images.githubusercontent.com/13059545/207921994-61a06897-0da6-492d-8995-c3e3fb84cb6d.png)

But indeed you are correct: the execute permission was still missing in GitHub. This used to work properly in WSL. Clearly not anymore.  

I've made the change on Ubuntu in a VM and confirmed it is now working with a fresh clone:

![image](https://user-images.githubusercontent.com/13059545/207920460-f268343b-a022-4f4a-b33e-4dd0480df31a.png)

gojimmypi commented:
Heads up I added the [include.am](https://github.com/gojimmypi/wolfssl/blob/ConfigCheck/config_check/include.am) to [Makefile.am](https://github.com/gojimmypi/wolfssl/blob/1bb40a77fd4dc183ad3cb6804548cf4102fecb0e/Makefile.am#L178).

I also don't know why I was [previously](https://github.com/gojimmypi/wolfssl/blob/da2e0b33f325cf7287930672b14a21602fa78a39/config_check/refresh.sh#L79)  _moving_ the the `options.h`. I changed that to a [copy](https://github.com/gojimmypi/wolfssl/blob/1bb40a77fd4dc183ad3cb6804548cf4102fecb0e/config_check/refresh.sh#L123).

jpbland1 reviewed (approved):
looks good, refresh is executable and an error message prints out when I remove configure from cmd.txt

gojimmypi commented:
jenkins retest this please

jpbland1 commented:
github says this branch is behind master, maybe merge the new changes and see if that fixes the build problem

dgarske reviewed (changes requested):
Please move this into `scripts/config_check`.

On Mac this doesn't work:

```
% ./refresh.sh
This WOLFSSL_REPO = /Users/davidgarske/GitHub/wolfssl
WOLFSSL_FILE_ROOT = /Users/davidgarske/GitHub/wolfssl/config_check
CMD File= /Users/davidgarske/GitHub/wolfssl/config_check/cmd.txt

tee: illegal option -- -
usage: tee [-ai] [file ...]

tee: illegal option -- -
usage: tee [-ai] [file ...]
tee: illegal option -- -
usage: tee [-ai] [file ...]
Error running command /Users/davidgarske/GitHub/wolfssl/config_check/cmd.txt
```

Mac uses `-a` for append `-a      Append the output to the files rather than overwriting them.`. `usage: tee [-ai] [file ...]`

gojimmypi commented:
Replaced `tee --append` with `tee -a`. I don't have a Mac so I was unable to test on a Mac. Verified on WSL and real Ubntun VM.

Renamed the refresh script to `config_check.sh` and moved into `scripts` directory.  

Now that I re-read [that](https://github.com/wolfSSL/wolfssl/pull/5876#pullrequestreview-1228427400), @dgarske perhaps you meant literally the `scripts/config_check` directory? Not a problem if you'd like to isolate it all in a new subdirectory of `scripts`.

dgarske reviewed (changes requested):
Hey Jim,

Please move these files (named as they were) into a new directory `scripts/config_check/`.

The fix for `tee -a` works on my Mac.

Thanks

gojimmypi commented:
@dgarske - done; moved into [scripts/config_check on this PR ConfigCheck branch](https://github.com/gojimmypi/wolfssl/tree/ConfigCheck/scripts/config_check).

Should I try to squash commits during or after the PR review process? There are some rather petty ones cluttering up the history. Let me know the best practice for this.

dgarske reviewed (changes requested):
A couple minor typos in the README.md.

Yes please squash to a single commit. There is a way to do it on GitHub merge but it looses some of the author/approver history.

gojimmypi commented:
performed a

```
git reset $(git merge-base master $(git branch --show-current))
git push --force
```

to squash commits. Re-added config files & reopened this PR.

gojimmypi commented:
Reopening this PR after using [script to squash commits](https://gist.github.com/gojimmypi/669a726595b9b0272874f6132279d95b).

test with:
```
# prep 
git clone https://github.com/gojimmypi/wolfssl.git wolfssl_ConfigCheck_PR_Test --depth 1 --branch ConfigCheck
cd wolfssl_ConfigCheck_PR_Test 
./autogen.sh
cd scripts/config_check

# test
./refresh.sh
```

jpbland1 commented on scripts/config_check/README.md:
```diff
@@ -0,0 +1,118 @@
+
+#### `WOLFSSL_CMD_FILE`
+
+This the `.configure` command to edit. This is typically the location of the [cmd.txt](./cmd.txt) file
```
should be `This is the` not `This the`

gojimmypi replied:
ah yes, I thought I fixed this, but there was another. good catch. thank you!

gojimmypi commented on scripts/config_check/README.md:
```diff
@@ -0,0 +1,96 @@
+
+## Installation
+
+Place the  [refresh.sh](./refresh.sh) and [cmd.txt](./cmd.txt) files in a directory. Currently
```
It wasn't, but I added some verbiage that if the user wants to run it from someplace else...

gojimmypi commented on scripts/config_check/README.md:
```diff
@@ -0,0 +1,96 @@
+## Installation
+
+Place the  [refresh.sh](./refresh.sh) and [cmd.txt](./cmd.txt) files in a directory. Currently
+they are expected to be in a subdirectory of the wolfSSL source code, in this case called `scripts/config_check_`.
```
darn auto-add of double chars in visual studio. same happens with quotes. fixed

gojimmypi commented on scripts/config_check/README.md:
````diff
@@ -0,0 +1,96 @@
+Run the `./refresh.sh` command from the `wolfssl/scripts/config` directory:
+
+```
+cd wolfssl/scripts/config
````
fixed

gojimmypi commented on scripts/config_check/README.md:
```diff
@@ -0,0 +1,96 @@
+
+## Usage
+
+Run the `./refresh.sh` command from the `wolfssl/scripts/config` directory:
```
fixed

gojimmypi commented:
updated readme, [resync with upstream and squashed commits](https://gist.github.com/gojimmypi/669a726595b9b0272874f6132279d95b), reopened for review.

dgarske commented on scripts/config_check/refresh.sh:
```diff
@@ -0,0 +1,142 @@
+#!/bin/bash
```
This needs `+x` bit set. Your squash pushes are very odd. The keep closing and reopening and generating multiple notifications. Just add a single normal commit next time.

gojimmypi replied:
huh. not even touching that file in Windows after setting the `-x` in a VM seems to clear that bit upon re-adding files in a commit in WSL.

gojimmypi replied:
It seems that when I reset my branch and force push prior to recommitting squashed changes, GitHub sees that there are no changes in my branch and thus automatically sets the PR to "merged" and closes it. Not at all desired nor intuitive.

gojimmypi replied:
As I think if this more, I agree. I'll add this text to the README.md for the output files:

These files are all included in the GitHub `.ignore` file in the root of the repository. 
Edit that file or redirect the output to a different location if you'd like to track changes.

gojimmypi replied:
regarding the `+x` that keeps getting lost in WSL: [stackoverflow](https://stackoverflow.com/questions/68281828/how-can-i-keep-the-execution-permissions-when-cloning-a-repository-in-windows) to the rescue (see near the end _"Manually updating modes when your file system does not cooperate_":  

```
 git add refresh.sh --chmod=+x
```

dgarske commented on scripts/config_check/Disabled-Features.txt:
```diff
@@ -0,0 +1 @@
+After running the refresh.sh command, this file will contain a list of disabled wolfSSL features.
```
My concern is this is a tracked file, so after running the script this show as modified. I think it would be better to not track and have it in the `.gitignore`.

gojimmypi replied:
interesting feedback; I was specifically interested in "_what changed since I last ran the configure refresh_".

dgarske replied:
Then just save off a .old version.

gojimmypi replied:
added to `.gitignore`

dgarske commented on scripts/config_check/Enabled-Features.txt:
```diff
@@ -0,0 +1 @@
+After running the refresh.sh command, this file will contain a list of enabled wolfSSL features.
```
My concern is this is a tracked file, so after running the script this show as modified. I think it would be better to not track and have it in the `.gitignore`.

gojimmypi replied:
added to `.gitignore`

dgarske commented on scripts/config_check/options.h:
```diff
@@ -0,0 +1 @@
+/* After running the refresh.sh command, this file will contain a copy of the generated options.h file. */
```
My concern is this is a tracked file, so after running the script this show as modified. I think it would be better to not track and have it in the `.gitignore`.

gojimmypi replied:
added to `.gitignore`

dgarske commented on scripts/config_check/output.txt:
```diff
@@ -0,0 +1 @@
+After running the refresh.sh command, this file will contain the output history.
```
My concern is this is a tracked file, so after running the script this show as modified. I think it would be better to not track and have it in the `.gitignore`.

gojimmypi replied:
added to `.gitignore`

kaleb-himes commented:
Initial thoughts on testing this out and reading the output:

1) Some of the comments seem hard-coded and if/when something is fixed will become obsolete
    - Example from cmd.txt -> ` --disable-md4                 # !! this does not actually disable MD4 !! use NO_MD4`
    Observation: Down the road items may get fixed or changed... are the comments in cmd.txt going to automatically update or will someone have to manually maintain this long term? (Q: What is the potential technical debt associated with this solution?)
    
2) I've often needed to review things in the log which this solution seems to be getting at but my approach was to simply echo it to a file for review `./configure <options> > kaleb-config.log 2>&1` (NOTE: `2>&1` echo's stderr to stdout and all of stdout then gets directed to the file so everything is there when you need to review it)
    Q: Can you shed some light on what it is that this solution offers above and beyond a simple approach like the above?
        Q2: Is that additional offering worth the technical debt I suspect will result to maintain this and keep it up to date?

dgarske commented on scripts/include.am:
```diff
@@ -126,3 +126,8 @@ endif
 EXTRA_DIST += scripts/user_settings_asm.sh
+
+EXTRA_DIST += \
+        scripts/config_check/README.md  \
```
`help.txt` is missing. If this is just output from `./configure --help` then you should just document how to use that and not make a copy of build options that needs updated.

gojimmypi replied:
I've removed the `help.txt` file and added a comment in the README.

kaleb-himes reviewed (commented):
Comment posted above.

gojimmypi commented:
@kaleb-himes I created this script mainly to have the ability to have a fully commented list of parameters to feed to the `./configure` command. Sure, there are a variety of ways to compare the output. I personally like the separate lists of enabled and disabled features in separate files. In particular, I liked having the configuration history with my commits when it came time for memory footprint analysis.

Note that in my case, I was using it for an embedded project. So I was manually copying some generated settings from the `./configure` to the project `user_settings.h` file. Keeping the  `./configfure` parameters and result coordinated with my `user_settings.h` and code history was useful at the time.

Good point on the `cmd.txt` file; Your are correct that the sample as-is may not be appropriate, and in particular the comments. Do you have a suggestion for a good default configuration? Otherwise I can prune to something smaller with more obvious sample comments.

gojimmypi commented:
I appreciate the review and attention to detail. I've made some changes as suggested.

gojimmypi commented:
Jenkins retest this please

gojimmypi commented:
There are probably better tools available. I'm closing this lingering draft PR as I likely won't submit it in the near future.
